src/Gaffer/ContextQuery.cpp

```diff
@@ -0,0 +1,342 @@
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "Gaffer/ContextQuery.h"
+
+#include "Gaffer/PlugAlgo.h"
+#include "Gaffer/TypedObjectPlug.h"
+
+#include "IECore/NullObject.h"
+
+#include "boost/bind.hpp"
+
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+
+namespace
+{
+
+const size_t g_existsPlugIndex = 0;
+const size_t g_valuePlugIndex = 1;
+
+/// \todo: Can the next function be move to somewhere to be shared with `AttributeQuery`?
+
+const Gaffer::ValuePlug *correspondingPlug(
+	const Gaffer::ValuePlug *parent,
+	const Gaffer::ValuePlug *child,
+	const Gaffer::ValuePlug *other
+)
+{
+	boost::container::small_vector< const Gaffer::ValuePlug*, 4 > path;
+
+	const Gaffer::ValuePlug *plug = child;
+
+	while( plug != parent )
+	{
+		path.push_back( plug );
+		plug = plug->parent< Gaffer::ValuePlug >();
+	}
+
+	plug = other;
+
+	while( ! path.empty() )
+	{
+		plug = plug->getChild< Gaffer::ValuePlug >( path.back()->getName() );
+		path.pop_back();
+	}
+
+	return plug;
+}
+
+void addChildPlugsToAffectedOutputs( const Gaffer::Plug* plug, Gaffer::DependencyNode::AffectedPlugsContainer& outputs )
+{
+	if( plug->children().empty() )
+	{
+		outputs.push_back( plug );
+	}
+	else
+	{
+		for( const Gaffer::PlugPtr& child : Gaffer::Plug::OutputRange( *plug ) )
+		{
+			addChildPlugsToAffectedOutputs( child.get(), outputs );
+		}
+	}
+}
+
+/// Returns the child of `parentPlug` that has `descendantPlug` as a descendant or is descendantPlug.
+/// Throws an Exception if the `descendantPlug` is not a descendant of `parentPlug`.
+const Gaffer::ValuePlug *getChildWithDescendant( const Gaffer::Plug *parentPlug, const Gaffer::ValuePlug *descendantPlug )
+{
+	const ValuePlug *p = descendantPlug;
+	while( p )
+	{
+		if( p->parent() == parentPlug )
+		{
+			return p;
+		}
+		p = p->parent<ValuePlug>();
+	}
+
+	throw IECore::Exception( "ContextQuery : Plug not in hierarchy." );
+}
+
+}  // namespace
+
+namespace Gaffer
+{
+
+GAFFER_NODE_DEFINE_TYPE( ContextQuery );
+
+size_t ContextQuery::g_firstPlugIndex = 0;
+
+ContextQuery::ContextQuery( const std::string &name ) : Gaffer::ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new ArrayPlug( "queries" ) );
+	addChild( new ArrayPlug( "out", Plug::Direction::Out ) );
```

Comment:
I think we need to disable the `resizeWhenInputsChange` behaviour of ArrayPlug here (via the constructor argument), so that `addQuery()` is in sole control of queries. I _think_ this will fix the following bug :

1. Make ContextQuery.
2. Add a string query via the UI.
3. Right click on the `name` widget for the query, and choose `Create Spreadsheet` or `Create Python Expression`.
4. Note that a new query has appeared, along with this message in the shell : `ERROR : gui.scripts.ScriptNode.ContextQuery::affects() : ContextQuery : "outPlug" is missing`.

I believe the same applies to ShaderQuery.

Reply:
This one is worth doing for ShaderQuery on `0.61_maintenance` @ericmehl.